### PR TITLE
fix: Invalid association between visitors and contacts.

### PR DESCRIPTION
**Pull request by pierre-lehnen-rc**

### CODE REVIEW

File: apps/meteor/client/views/omnichannel/contactInfo/tabs/ContactInfoChannels/ContactInfoChannels.tsx [hunk 1]
```diff
@@ -46,7 +46,12 @@ const ContactInfoChannels = ({ contactId }: ContactInfoChannelsProps) => {
 							overscan={25}
 							data={data?.channels}
 							components={{ Scroller: VirtuosoScrollbars }}
-							itemContent={(_, data) => <ContactInfoChannelsItem key={data.visitorId} {...data} />}
+							itemContent={(_, data) => (
+								<ContactInfoChannelsItem
+									key={`id=${data.visitor.visitorId};type=${data.visitor.source.type};id=${data.visitor.source.id || ''}`}
```
Comment (dougfabris) [inline apps/meteor/client/views/omnichannel/contactInfo/tabs/ContactInfoChannels/ContactInfoChannels.tsx:51]: let's use the index here for sake of sanity

**Review by dougfabris — CHANGES_REQUESTED** [review]

File: apps/meteor/app/livechat/server/lib/contacts/registerGuestData.ts [hunk 2]
```diff
@@ -35,5 +29,9 @@ export async function registerGuestData(
 		return;
 	}
 
-	await ContactMerger.mergeFieldsIntoContact(fields, contact, contact.unknown ? 'overwrite' : 'conflict');
+	// If a visitor was updated who already had contacts, load up the contacts and update that information as well
+	const contacts = await LivechatContacts.findAllByVisitorId(visitor._id).toArray();
+	for await (const contact of contacts) {
+		await ContactMerger.mergeFieldsIntoContact(fields, contact, contact.unknown ? 'overwrite' : 'conflict');
```
Comment (tapiarafael) [inline apps/meteor/app/livechat/server/lib/contacts/registerGuestData.ts:35]: Will this fix [SCI-136](https://rocketchat.atlassian.net/browse/SCI-136)?

File: apps/meteor/ee/app/livechat-enterprise/server/api/contacts.ts [hunk 3]
```diff
@@ -9,17 +11,15 @@ const ajv = new Ajv({
...
 const blockContactSchema = {
 	type: 'object',
 	properties: {
-		visitorId: {
-			type: 'string',
-		},
+		visitor: ContactVisitorAssociationSchema,
 	},
-	required: ['visitorId'],
+	required: ['visitorId', 'source'],
```
Comment (tapiarafael) [inline apps/meteor/ee/app/livechat-enterprise/server/api/contacts.ts:22]:
```suggestion
	required: ['visitor'],
```

**Review by tapiarafael — CHANGES_REQUESTED** [review]

File: apps/meteor/ee/app/livechat-enterprise/server/api/contacts.ts (hunk 3 above)
Comment (pierre-lehnen-rc) [reply apps/meteor/ee/app/livechat-enterprise/server/api/contacts.ts:22]: no, this is right; you can't block a channel without specifying both visitor and source.

Comment (pierre-lehnen-rc) [reply apps/meteor/ee/app/livechat-enterprise/server/api/contacts.ts:22]: OHH I got the issue now.

File: apps/meteor/app/livechat/server/lib/contacts/registerGuestData.ts (hunk 2 above)
Comment (pierre-lehnen-rc) [reply apps/meteor/app/livechat/server/lib/contacts/registerGuestData.ts:35]: Probably not by itself, as this was already being done before.

**Review by tapiarafael — APPROVED** [review]

**Pull request merged**